.github/workflows/tests.yml

```diff
@@ -0,0 +1,249 @@
+        uses: actions/checkout@v4
+
+      - name: Install dependencies
+        run: |
+          sudo apt install tox
+
+      - name: Run coverage
+        run: |
+          tox -e cover
+
+  build:
+    name: Build charm
+    needs:
+      - lint
+      - unit-py38
+      - unit-py310-py311
+      - coverage
+    runs-on: ubuntu-22.04
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - name: Install dependencies
+        run: |
+          sudo apt install tox
+          sudo snap install charmcraft --channel 2.x/stable --classic
+          sudo snap install lxd --channel 5.21/stable
+          sudo snap set lxd daemon.group=adm
+          sudo lxd init --auto
+
+      - name: Clear FORWARD firewall rules
+        run: |
+          # Docker can inject rules causing firewall conflicts
+          sudo iptables -P FORWARD ACCEPT  || true
+          sudo ip6tables -P FORWARD ACCEPT || true
+          sudo iptables -F FORWARD  || true
+          sudo ip6tables -F FORWARD || true
+
+      - name: Build charm
+        run: |
+          tox -e build
+
+      - name: Store charm as artifact
+        uses: actions/upload-artifact@v4
+        with:
+          name: ovn-central-charm
+          path: ./ovn-central_ubuntu-22.04-amd64.charm
+          retention-days: 7
+
+  parse_tags:
+    name: Parse tags in PR commits
+    runs-on: ubuntu-latest
+    permissions:
+      pull-requests: read
+    outputs:
+      zaza_pr: ${{ steps.find-zaza-pr.outputs.zaza_pr }}
+
+    steps:
+      - name: Get PR commits
+        id: 'get-pr-commits'
+        uses: tim-actions/get-pr-commits@198af03565609bb4ed924d1260247b4881f09e7d
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Find requested functional test PR
+        id: 'find-zaza-pr'
+        run: |
+          # Note(mkalcok): The command below cleanely transfers data from Github
+          #                variable to bash variable without interpreting curly
+          #                braces or losing quotations.
+          COMMIT_DATA=$(
+          cat <<'EOF'
+          ${{ steps.get-pr-commits.outputs.commits }}
+          EOF
+          )
+
+          ZAZA_TAG="Func-test-pr: "
+          FUNC_PR=$(jq -r '.[].commit.message' <<< "$COMMIT_DATA" | grep -i "^${ZAZA_TAG}" | tail -1)
+          PR_URL=$(echo $FUNC_PR | sed -e "s/^$ZAZA_TAG//I")
+
+          echo "zaza_pr=$PR_URL" >> "$GITHUB_OUTPUT"
+
+  functest:
+    name: Functional Tests
+    runs-on: [self-hosted, linux, X64, large, focal]
+    needs:
+      - build
+      - parse_tags
+    strategy:
+      matrix:
+        bundle: ["jammy-caracal"]
+    env:
+      DEBIAN_FRONTEND: noninteractive
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      # ovn-central branch stable/24.03 is locked to run functional tests
+      # with Python 3.8
+      - name: Set up Python 3.8
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.8"
```

Comment:
I have a feeling this may be a past sin / side effect of battling the old CI.

I see you have made good effort in making this work to minimize the amount of change needed to the stable branch, so let's keep it.

If this causes issues for the CI in the future I'd be interested in us reverting the change that locked the branch to Python 3.8 testing and then we can also remove these bits.

Reply:
Thanks for noticing this. It is indeed an oversight that I forgot to remove. Since the whole job runs on focal, which has python3.8 natively, this step has basically no effect. I'll propose the removal